File: solutions/1137. N-th Tribonacci Number/1137.py
# Time complexity: O(n)
# Space complexity: O(1)
class Solution:
  def tribonacci(self, n: int) -> int:
    if n <= 2:
      return 0 if n == 0 else 1
    
    a, b, c = 0, 1, 1
    for i in range(3, n + 1):
      t = a + b + c
      a, b, c = b, c, t

    return c   

File: solutions/1137. N-th Tribonacci Number/1137.cpp
// Time complexity: O(n)
// Space complexity: O(1)
class Solution {
 public:
  int tribonacci(const int& n) {
    if (n <= 2)
      return n == 0 ? 0 : 1;

    int a = 0, b = 1, c = 1;
    for (int i  = 3; i <= n; ++i) {
      const int t =  a + b + c;
      a = b;
      b = c;
      c = t;
    }

    return c;
  }
};